src/types/redis_bitmap.cc

```diff
@@ -538,6 +546,320 @@ rocksdb::Status Bitmap::BitOp(BitOpFlags op_flag, const std::string &op_name, co
+  }
+  return rocksdb::Status::OK();
+}
+
+}  // namespace
+
+template <bool ReadOnly>
+rocksdb::Status Bitmap::bitfield(const Slice &user_key, const std::vector<BitfieldOperation> &ops,
+                                 std::vector<std::optional<BitfieldValue>> *rets) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  std::optional<LockGuard> guard;
+  if constexpr (!ReadOnly) {
+    guard = LockGuard(storage_->GetLockManager(), ns_key);
+  }
+
+  BitmapMetadata metadata;
+  std::string raw_value;
+  auto s = GetMetadata(ns_key, &metadata, &raw_value);
+  if (!s.ok() && !s.IsNotFound()) {
+    return s;
+  }
+
+  if (metadata.Type() == RedisType::kRedisString) {
+    if constexpr (ReadOnly) {
+      s = BitmapString::BitfieldReadOnly(ns_key, raw_value, ops, rets);
+    } else {
+      s = BitmapString(storage_, namespace_).Bitfield(ns_key, &raw_value, ops, rets);
+    }
+    return s;
+  }
+
+  if (metadata.Type() != RedisType::kRedisBitmap) {
+    return rocksdb::Status::InvalidArgument("The value is not a bitmap or string.");
+  }
+
+  // We firstly do the bitfield operation by fetching segments into memory.
+  // Use SegmentCacheStore to record dirty segments. (if not read-only mode)
+  SegmentCacheStore cache(storage_, metadata_cf_handle_, ns_key, metadata);
+  runBitfieldOperationsWithCache<ReadOnly>(cache, ops, rets);
+
+  if constexpr (ReadOnly) {
+    return rocksdb::Status::OK();
+  }
+
+  // Write changes into storage.
+  auto batch = storage_->GetWriteBatchBase();
+  bitfieldWriteAheadLog(batch, ops);
+  cache.BatchForFlush(batch);
+  return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+template <bool ReadOnly>
+rocksdb::Status Bitmap::runBitfieldOperationsWithCache(SegmentCacheStore &cache,
+                                                       const std::vector<BitfieldOperation> &ops,
+                                                       std::vector<std::optional<BitfieldValue>> *rets) {
+  ArrayBitfieldBitmap tmp_bitfield;
+  for (BitfieldOperation op : ops) {
+    // found all bytes that contents the bitfield.
+    uint32_t first_byte = op.offset / 8;
+    uint32_t last_bytes = (op.offset + op.encoding.Bits() - 1) / 8 + 1;
+    uint32_t bytes = last_bytes - first_byte;
+
+    ArrayBitfieldBitmap bitfield;
+    auto segment_status = CopySegmentsBytesToBitfield(cache, first_byte, bytes, &bitfield);
+    if (!segment_status.ok()) {
+      return segment_status;
+    }
+
+    // Covert the bitfield from a buffer to an integer.
+    uint64_t unsigned_old_value = 0;
+    auto s = GetBitfieldInteger(bitfield, op.offset, op.encoding, &unsigned_old_value);
+    if (!s.ok()) {
+      return s;
+    }
+
+    if constexpr (ReadOnly) {
+      rets->emplace_back() = {op.encoding, unsigned_old_value};
+      continue;
+    }
+
+    auto &ret = rets->emplace_back();
+    uint64_t unsigned_new_value = 0;
+    // BitfieldOp failed only when the length or bits not illegal.
```

Comment:
There is a typo. Should be `BitfieldOp failed only when the length or bits illegal.`